### PR TITLE
Fix SIGINT/SIGTERM [V4]

### PR DESCRIPTION
**Pull request by apahim**
v4:
- Selftests:
  - Refactor to reuse the common methods.
  - Use `stderr=subprocess.STDOUT` 

v3: #2140 
- Change the approach to fix this issue. Previously, I was trying to handle the signals from the main process all the way to the test sub-process, through all the layers in between. That was proven hand to control and with a big number of corner cases. Now, the approach is rather simple:
  - When a SIGINT in received by the main process, propagate the SIGINT for all the children.
  - When a SIGTERM is received by the main process, send a SIGKILL for all the children.
- Tests where rewritten to cover all the scenarios we have so far.

v2: #2137 
- Fix message on SIGINT (new commit)
- Improve SIGINT selftest.
- New commit to fix SIGINT/SIGTERM behavior on avocado.runner and process.Subprocess. See commit message for details.

v1: #2130

### CODE REVIEW

**Comment by clebergnu** [comment]
[BUG: handling of SIGINT on test runner freezes it](https://trello.com/c/5RPZPH4F/1068-bug-handling-of-sigint-on-test-runner-freezes-it)

File: selftests/functional/test_interrupt.py [hunk 1]
```diff
@@ -67,64 +77,87 @@ def test_badly_behaved(self):
...
         bad_test.save()
-
+        self.test_module = bad_test.path
         os.chdir(basedir)
-        cmd_line = ('%s run --sysinfo=off --job-results-dir %s '
-                    '%s %s %s' % (AVOCADO, self.tmpdir, bad_test.path,
-                                  bad_test.path, bad_test.path))
-        proc = aexpect.Expect(command=cmd_line, linesep='')
-        proc.read_until_last_line_matches(os.path.basename(bad_test.path))
-        proc.sendline('\x03')
```
Comment (clebergnu) [inline selftests/functional/test_interrupt.py]: Right, it's fine to replace a key that will be translated as SIGINT by the signal itself.  The current difference in treatment is actually one of the points of the original card hoped to resolve.

File: selftests/functional/test_interrupt.py [hunk 2]
```diff
@@ -67,64 +77,87 @@ def test_badly_behaved(self):
...
-                except psutil.NoSuchProcess:
-                    pass
-
-            return len(bad_test_processes) == 0
-
-        wait.wait_for(wait_until_no_badtest, timeout=2)
+        if not wait.wait_for(self._no_test_in_process_table, timeout=2):
+            self.fail('Avocado left processes behind.')
+
+        output = self.proc.communicate()[0]
```
Comment (clebergnu) [inline selftests/functional/test_interrupt.py]: This works, but I think it passes the wrong message to readers (it happened to me, at least).  Popen.communicate has the side effect of returning stdout (and stderr), but it's intended to send data to the process' stdin.  How about:

```diff
diff --git a/selftests/functional/test_interrupt.py b/selftests/functional/test_interrupt.py
index 4325b1a3..e9d9ce31 100644
--- a/selftests/functional/test_interrupt.py
+++ b/selftests/functional/test_interrupt.py
@@ -106,7 +106,7 @@ class InterruptTest(unittest.TestCase):
         if not wait.wait_for(self._no_test_in_process_table, timeout=2):
             self.fail('Avocado left processes behind.')
 
-        output = self.proc.communicate()[0]
+        output = self.proc.stdout.read()
         # Make sure the Interrupted requested sentence is there
         self.assertIn('Interrupt requested. Waiting 2 seconds for test to '
                       'finish (ignoring new Ctrl+C until then)', output)
@@ -150,9 +150,8 @@ class InterruptTest(unittest.TestCase):
         if not wait.wait_for(self._no_test_in_process_table, timeout=2):
             self.fail('Avocado left processes behind.')
 
-        output = self.proc.communicate()[0]
         # Make sure the Interrupted test sentence is there
-        self.assertIn('Terminated\n', output)
+        self.assertIn('Terminated\n', self.proc.stdout.read())
 
     @unittest.skipIf(int(os.environ.get("AVOCADO_CHECK_LEVEL", 0)) < 1,
                      "Skipping test that take a long time to run, are "
@@ -191,7 +190,7 @@ class InterruptTest(unittest.TestCase):
         if not wait.wait_for(self._no_test_in_process_table, timeout=2):
             self.fail('Avocado left processes behind.')
 
-        output = self.proc.communicate()[0]
+        output = self.proc.stdout.read()
         # Make sure the Interrupted requested sentence is there
         self.assertIn('Interrupt requested. Waiting 2 seconds for test to '
                       'finish (ignoring new Ctrl+C until then)', output)
@@ -235,9 +234,8 @@ class InterruptTest(unittest.TestCase):
         if not wait.wait_for(self._no_test_in_process_table, timeout=2):
             self.fail('Avocado left processes behind.')
 
-        output = self.proc.communicate()[0]
         # Make sure the Interrupted test sentence is there
-        self.assertIn('Terminated\n', output)
+        self.assertIn('Terminated\n', self.proc.stdout.read())
 
     def tearDown(self):
         shutil.rmtree(self.tmpdir)

```

File: selftests/functional/test_interrupt.py [hunk 3]
```diff
@@ -134,59 +167,121 @@ def test_well_behaved(self):
...
+        self.proc = subprocess.Popen(cmd.split(),
+                                     stdout=subprocess.PIPE,
+                                     stderr=subprocess.STDOUT)
+
+        if not wait.wait_for(self._has_children, timeout=2):
+            process.kill_process_tree(self.proc.pid)
+            self.fail('Avocado did not start the test process.')
+
+        # This test will not ignore SIGINT, so it will
+        # terminate right away.
```
Comment (clebergnu) [inline selftests/functional/test_interrupt.py]: This code is fine, I just had to say that affirming the process *will* terminate right away "signals" a lot of confidence! (pun intended).  It *should*, though, and this is what we're testing.
